docs/develop/typescript/message-passing.mdx

```diff
@@ -27,123 +27,227 @@ tags:
 
-export async function yourWorkflow() {
-  const groups = new Map<string, Set<string>>();
 
-  setHandler(joinSignal, ({ userId, groupId }: JoinInput) => {
-    const group = groups.get(groupId);
-    if (group) {
-      group.add(userId);
-    } else {
-      groups.set(groupId, new Set([userId]));
+
+### Update handlers and validators {#updates}
+
+An [Update](/encyclopedia/workflow-message-passing#sending-updates) is a trackable synchronous request sent to a running Workflow Execution.
+It can change the Workflow state, control its flow, and return a result.
+The sender must wait until the Worker accepts or rejects the Update.
+The sender may wait further to receive a returned value or an exception if something goes wrong:
+
+```typescript
+// 👉 Use the object returned by defineUpdate to set the Update handler in
+// Workflow code, and to send Updates from Client code.
+export const setLanguage = wf.defineUpdate<Language, [Language]>('setLanguage');
+
+export async function greetingWorkflow(): Promise<string> {
+  const greetings: Partial<Record<Language, string>> = {
+    [Language.CHINESE]: '你好，世界',
+    [Language.ENGLISH]: 'Hello, world',
+  };
+
+  let language = Language.ENGLISH;
+
+  wf.setHandler(setLanguage, (newLanguage: Language) => {
+    // 👉 An Update handler can mutate the Workflow state and return a value.
+    if (!(newLanguage in greetings)) {
+      throw new Error(`${newLanguage} is not supported`);
     }
+    const previousLanguage = language;
+    language = newLanguage;
+    return previousLanguage;
   });
+
+  ...
 }
 ```
 
-The Workflow dynamically reacts to external events without interrupting its execution flow.
-There can be only one handler for any Signal type.
-Calling `setHandler` multiple times for the same Signal type overwrites the existing handler.
-For example, if you first set Handler A for a given Signal, and then later in the Workflow Definition set Handler B for the same Signal, then only Handler B will be invoked upon receiving this Signal.
+- `setHandler` can take `UpdateHandlerOptions` (such as `validator`, `description` and `unfinishedPolicy`) as described in the API reference docs for [`workflow.setHandler`](https://typescript.temporal.io/api/namespaces/workflow#sethandler).
+
+- About validators:
+  - Use validators to reject an Update before it is written to History.
+    Validators are always optional.
+    If you don't need to reject Updates, you don't need a validator.
+  - To set a validator, pass the validator function in `UpdateHandlerOptions` when calling [`workflow.setHandler`](https://typescript.temporal.io/api/namespaces/workflow#sethandler).
+    The validator must be a non-async function that accepts the same argument types as the handler and returns `void`.
+
+- Accepting and rejecting Updates with validators:
+  - To reject an Update, throw an error of any type in the validator.
```

Comment:
@dandavison I don't remember the details from the top of my head… Is it ok to throw an `ApplicationFailure` from a validator? i.e. would that result in the same behavior as throwing, let's say, an `Error`?

Reply:
I just checked and `ApplicationFailure` behaves the same as `Error`.